src/System.IO.Compression.Brotli/System/IO/Compression/Brotli/BrotliPrimitives.cs

```diff
@@ -16,41 +17,67 @@ public struct BrotliPrimitives
         const int defQuality = 11;
         const int defLgWin = 24;
 
-        public static bool Compress(ReadOnlySpan<byte> source, Span<byte> destination, out nuint bytesConsumed, out nuint bytesWritten)
+        public static int GetMaximumCompressedSize(int input_size)
         {
-            return Compress(source, destination, out bytesConsumed, out bytesWritten, defQuality, defLgWin);
+            if (input_size == 0) return 1;
+            int num_large_blocks = input_size >> 24;
+            int tail = input_size & 0xFFFFFF;
+            int tail_overhead = (tail > (1 << 20)) ? 4 : 3;
+            int overhead = 2 + (4 * num_large_blocks) + tail_overhead + 1;
+            int result = input_size + overhead;
+            return (result < input_size) ? input_size : result;
         }
 
-        public static bool Compress(ReadOnlySpan<byte> source, Span<byte> destination, out nuint bytesConsumed, out nuint bytesWritten, int quality, int lgwin)
+        public static bool Compress(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesConsumed, out int bytesWritten, int quality=defQuality, int lgwin=defLgWin, BrotliEncoderMode encMode=BrotliEncoderMode.Generic)
         {
+            if (quality > defQuality || quality <= 0) throw new System.ArgumentOutOfRangeException(BrotliEx.WrongQuality);
+            if (lgwin > defLgWin || lgwin <= 0) throw new System.ArgumentOutOfRangeException(BrotliEx.WrongWindowSize);
+            bytesConsumed = 0;
+            bytesWritten = 0;
             unsafe
             {
-                bytesConsumed = (nuint)source.Length;
-                bytesWritten = (nuint)0;
                 IntPtr bufIn, bufOut;
                 fixed (byte* inBytes = &source.DangerousGetPinnableReference())
                 fixed (byte* outBytes = &destination.DangerousGetPinnableReference())
                 {
                     bufIn = new IntPtr(inBytes);
                     bufOut = new IntPtr(outBytes);
-                    return BrotliNative.BrotliEncoderCompress(quality, lgwin, BrotliNative.BrotliEncoderMode.Generic, bytesConsumed, bufIn, ref bytesWritten, bufOut);
+                    nuint written = (nuint)destination.Length;
+                    nuint consumed = (nuint)source.Length;
+                    if (!BrotliNative.BrotliEncoderCompress(quality, lgwin, encMode, consumed, bufIn, ref written, bufOut))
+                    {
+                        bytesConsumed = bytesWritten = 0;
```

Comment:
Removing the 0 writes at the top then. They will be overwritten in the vast majority of the cases below here anyways.